scipio/src/channels/local_channel.rs

```diff
@@ -0,0 +1,586 @@
+/// [`LocalChannel`]: struct.LocalChannel.html
+/// [`StreamExt`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html
+pub struct LocalReceiver<T> {
+    channel: LocalChannel<T>,
+}
+
+#[derive(Debug)]
+struct State<T> {
+    capacity: ChannelCapacity,
+    channel: VecDeque<T>,
+    recv_waiters: Vec<Waker>,
+    send_waiters: Vec<Waker>,
+    receiver_alive: bool,
+    sender_alive: bool,
+}
+
+#[derive(Debug, Clone)]
+/// Allow data to be transmitted across two tasks in the same shard.
+///
+/// Asynchronous code is rarely useful if it is serialized. In practice,
+/// you will want to spawn asynchronous tasks or in the case of Scipio which
+/// support multi-queued execution have a dedicated [`TaskQueue`] for each group
+/// of tasks with their own priority.
+///
+/// The `LocalChannel` is an abstraction similar to the [`channel`] that can be
+/// used for this purpose
+///
+/// The communication between sender and receiver is broken when one of them goes
+/// out of scope. They however behave differently:
+///
+/// * The [`LocalReceiver`] never sees an error, as it is implemented as a stream
+///   interface compatible with [`StreamExt`]. When the sender is no longer available the receiver's call to [`next`] will return [`None`].
+/// * The [`LocalSender`] will return a [`BrokenPipe`] error if it tries to [`push`] into
+///   a channel that no longer has a receiver.
+///
+/// # Examples
+///
+/// ```
+/// use scipio::{LocalExecutor, Local, Latency, Shares};
+/// use scipio::channels::LocalChannel;
+/// use futures::stream::StreamExt;
+///
+/// let ex = LocalExecutor::make_default();
+/// ex.run(async move {
+///     let task_queue =
+///         Local::create_task_queue(
+///             Shares::default(), Latency::NotImportant, "example");
+///     
+///     let (sender, mut receiver) = LocalChannel::new_unbounded();
+///     let h = Local::local_into(async move {
+///         assert_eq!(receiver.next().await.unwrap(), 0);
+///     }, task_queue).unwrap().detach();
+///     sender.send(0);
+///     drop(sender);
+///     h.await;
+/// });
+/// ```
+/// [`TaskQueue`]: struct.TaskQueue.html
+/// [`LocalSender`]: struct.LocalSender.html
+/// [`LocalReceiver`]: struct.LocalReceiver.html
+/// [`push`]: struct.LocalSender.html#method.push
+/// [`None`]: https://doc.rust-lang.org/std/option/enum.Option.html#variant.None
+/// [`BrokenPipe`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.BrokenPipe
+/// [`channel`]: https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html
+/// [`next`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html#method.next
+/// [`StreamExt`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html
+pub struct LocalChannel<T> {
+    state: Rc<RefCell<State<T>>>,
+}
+
+impl<T> LocalChannel<T> {
+    fn new(capacity: ChannelCapacity) -> (LocalSender<T>, LocalReceiver<T>) {
+        let channel = match capacity {
+            ChannelCapacity::Unbounded => VecDeque::new(),
```

Comment:
Is there a good use case for unbounded channels?

Reply:
Non-DAG communication graphs (actor systems) need at least some channels to be unbounded, otherwise they deadlock: https://trio.discourse.group/t/sizing-the-channel-deadlock-freedom-vs-back-pressure/311

Reply:
More importantly, this is not an opinion that I would like Scipio to have.
Other crates for channels have this which indicates that some people have uses for unbounded channels.


Scipio should strive for being generic and making many use cases possible.

Reply:
It could have that opinion, similar to how Go has it. Unbounded queues are usually anathema for latency-sensitive solutions (unbounded queues do solve deadlocks in CSP systems, in a similar way that a Big Lock can solve data races).

Reply:
you are right about the latency issue but I would like scipio to be useful to build latency sensitive applications, but *not only* to build latency sensitive applications.

Sometimes the complexity of the application is elsewhere.